### ISSUE TITLE
Pypandoc is depedency of Pyspark = 2.2.2 and with todays change of python > 3 some of our some ancient pipelines started to break

### ISSUE THREAD

**Issue by shashanksinghfd** [NicklasTegner/pypandoc#246]
Heya guys,

Just looking for some suggestions from the community, to what they are doing. We have some ancient pipelines that are being migrated albeit slowly out of python 2 to python 3. Today's commit broke our build, and we were wondering if anyone else has the same issue or what they are doing as a stop-gap solution.

Commit Link : https://github.com/NicklasTegner/pypandoc/commit/9ee669801ec2b5837e95f412f3bce46c2f454a28

**Comment by NicklasTegner** [comment]
Hi.
In what way are they breaking. Do you have any error output in regards to PyPandoc?

**Comment by shashanksinghfd** [comment]
Thanks @NicklasTegner for replying.

**Issue**: Package u'pypandoc' requires a different Python: 2.7.17 not in '>=3,<4'
**Full traceback** :
```
ERROR: Command errored out with exit status 1:
--
  | command: /erie/_venvs/glue-2.7/bin/python2.7 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-rISGJs/pyspark/setup.py'"'"'; __file__='"'"'/tmp/pip-install-rISGJs/pyspark/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-9pydeC
  | cwd: /tmp/pip-install-rISGJs/pyspark/
  | Complete output (24 lines):
  | Could not import pypandoc - required to package PySpark
  | DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support pip 21.0 will remove support for this functionality.
  | ERROR: Package u'pypandoc' requires a different Python: 2.7.17 not in '>=3,<4'
  | Traceback (most recent call last):
  | File "<string>", line 1, in <module>
  | File "/tmp/pip-install-rISGJs/pyspark/setup.py", line 214, in <module>


```

**Comment by NicklasTegner** [comment]
Hi.
The recent fix as a result of #247  which fixed python 2.7 support should have fixed this as welll.
Please confirm if you still get the above error when using pypandoc 1.7.2

**Comment by shashanksinghfd** [comment]
@NicklasTegner , its downstream dependency of pyspark , so we can't directly change the version without forking pyspark itself. but thanks we might look into that too .

**Comment by jankatins** [comment]
@shashanksinghfd You can probably pin the pypandoc dependency in your deployment script. Seems this hasn't happen for some reason in pyspark itself, when it was still used :-( (e.g. this is the commit which rips out pypandoc from spark: https://github.com/apache/spark/commit/bda0669110fcfce5279221d2b9c2986a3cc7d15a)

As a workaround on your side, it would probably be better to install pypandoc in a pinned version before installing pyspark: the requirement would already be satisfied and pyspark could be installed. Given that pining dependencies in deployments is a best practice, this seems anyway reasonable.

This would have the additional advantage that the maintenance burden is not dropped onto an open source maintainer (= @NicklasTegner in this case). py27 is dead since a year, it's not sustainable to keep requiring open source tools to still test/develop against it.

Maybe another good thing would be to add an explicit python requirement in setup.py: `python_requires='>3.5.2',` (from https://stackoverflow.com/a/48777286). This should make pip (if it's a reasonable new version...) use the old packages which still support python 2.7.

**Comment by shashanksinghfd** [comment]
makes sense @jankatins : that's what we ended up going with. ( pinning pypandoc and corresponding pyspark version that has no issues )

thanks for the support @NicklasTegner @jankatins . Open source maintainers are heroes :)
Coffee/beer is on me if you guys find yourself in London/Edinburgh.

**Comment by NicklasTegner** [comment]
hi @shashanksinghfd 
The issues with python2 support (which if I read correctly is the issue here) should be resolved in pypandoc 1.7.2, which readded python2.7 as a supported version again.
Please let me know if it works :)

**Comment by shashanksinghfd** [comment]
thanks, we ended up updating the spark package which also solved the issue for us. I WIll close this ticket for now.
